cuqipy_fenics/geometry.py

```diff
@@ -398,3 +403,243 @@ def _build_basis(self):
+class _StepExpression(dl.UserExpression):
+    def __init__(self, degree, x_lim, y_lim=None):
+        self.x_lim = x_lim
+        self.y_lim = y_lim
+        super().__init__(degree=degree)
+
+    def eval(self, value, x):
+        if (
+            x[0] >= self.x_lim[0]
+            and x[0] < self.x_lim[1]
+            and (self.y_lim is None or x[1] >= self.y_lim[0])
+            and (self.y_lim is None or x[1] < self.y_lim[1])
+        ):
+            value[0] = 1
+        else:
+            value[0] = 0
+
+
+class StepExpansion(_WrappedGeometry):
+    """A geometry class that parameterizes finite element functions on a 1D or
+    2D domain with piecewise constant functions. In the 1D case, the domain is
+    divided into `num_steps_x` constant functions, each function support is of
+    length `L_x`/`num_steps_x`, where `L_x` is the length of the domain in the
+    x direction. In the 2D case, the domain is divided into `num_steps_x`
+    :math:`\\times` `num_steps_y` constant functions each has a support area of
+    `L_x`/`num_steps_x` x `L_y`/`num_steps_y`, where `L_y` is the length of the
+    domain in the y direction. The constant functions are arranged in a row-wise
+    order, from the bottom to top of the domain. That is, the first `num_steps_x`
+    functions are in the bottom row, the next `num_steps_x` functions are in the
+    next row, and so on. The step expansion is built on the given input geometry
+    that specifies the computational mesh.
+
+    Note: for accurate results, the underlying mesh should be structured in
+    which the number of nodes in in the x and y directions are multiples of
+    `num_steps_x` and `num_steps_y` respectively.
+
+
+    Parameters
+    -----------
+    geometry : cuqipy_fenics.geometry.Geometry
+        An input geometry on which the step expansion is built (the
+        geometry must have a mesh attribute)
+
+    num_steps_x: int
+        Number of step expansion terms in the x direction
+
+    num_steps_y: int, optional
+        Number of step expansion terms in the y direction,
+        only required for geometries with physical dimension 2
+
+
+    Example
+    -------
+    .. code-block:: python
+
+        import numpy as np
+        from cuqipy_fenics.geometry import StepExpansion, FEniCSContinuous
+        from cuqi.distribution import Gaussian
+        import dolfin as dl
+
+
+        mesh = dl.UnitSquareMesh(32,32)
+        V = dl.FunctionSpace(mesh, 'DG', 0)
+
+        G_FEM = FEniCSContinuous(V, labels=['$\\xi_1$', '$\\xi_2$'])
+        G_step = StepExpansion(G_FEM,
+                                    num_steps_x=4, num_steps_y=8)
+
+        x = Gaussian(mean=np.zeros(G_step.par_dim),
+                     cov=np.eye(G_step.par_dim),
+                     geometry=G_step)
+
+        samples = x.sample()
+        samples.plot()
+
+    """
+
+    def __init__(self, geometry, num_steps_x=8, num_steps_y=None):
+        # ensure num_steps_x is a positive integer
```

Comment:
A small thing. It seems clear from both the if statement and the value error what the code does. I would say that in this case the comment is not needed.

Reply:
good point, I thought to keep it though to be consistent with the comment in line 492 about steps_y. In that line the logic is a bit more involved. It is true that this comment repeats what is in the error message, but it could be that the code reader might start thinking about the "if" statement meaning before looking at the error message.